Implement statement descriptor sanitation

samiuelson opened the pull request:
Introducing `StatementDescriptor` class which performs sanitation of the original statement descriptor in `StatementDescriptor::value` getter.

<!-- Remember about a good descriptive title. -->

Closes: #7586

## Description
`StatementDescriptor` acts as a wrapper to raw statement descriptor text and applies a set of sanitation operations to prepare it for use with Stripe API. There are the following requirements applied:
 * Text must be max 22 chars long.
 * It must not contain the following characters: <, >, ', ".
 * Similarly to [iOS implementation](https://github.com/woocommerce/woocommerce-ios/blob/trunk/Hardware/Hardware/CardReader/StatementDescriptor.swift) illegal chars are replaced by "-" char.

## Testing
The crash was reproduced in case statement descriptor (which by default is made of a website's domain) was too long -exceeded 22 chars in length.

To test this correctly you'd need to reproduce the crash on `trunk`:
1. Connect to the store with a descriptor name length exceeding 22 chars or containing illegal chars. 
2. Observe the crash when collecting payment using card.

Next, verify that the crash is not happening on `issue-7586/statement-descriptor-improvement` branch.

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

samiuelson committed:
Implement statement descriptor sanitation

* Introduce `StatementDescriptor` class which performs sanitation of the original statement descriptor in `StatementDescriptor::value` getter.

samiuelson committed:
Add documentation

samiuelson committed:
Update RELEASE-NOTES.txt

pachlava reviewed (approved):
Thanks for working on this @samiuelson 👋 

My review consisted of making sure that the store with descriptor name being longer than 22 characters (the one which uncovered the original bug) can now be used for Stripe card payments, when the installable from this PR is used - and it's indeed possible now 👍

kidinov commented on RELEASE-NOTES.txt:
```diff
@@ -3,6 +3,7 @@
 -----
 - [**] Login: add support for Jetpack connection to sites that have Jetpack but don't have any connected accounts [https://github.com/woocommerce/woocommerce-android/pull/7545]
 - [*] [Internal] In-Person Payments: add UTM parameters to card reader purchase URLs to allow attribution [https://github.com/woocommerce/woocommerce-android/pull/7554]
+- [*] Fixed a bug caused by statement descriptor value too long or containing illegal chars [https://github.com/woocommerce/woocommerce-android/pull/7595]
```
It looks like this comes to the 10.10 release, not 10.9.

kidinov replied:
Sorry, 11.0 =)

kidinov commented on libs/cardreader/src/main/java/com/woocommerce/android/cardreader/payments/StatementDescriptor.kt:
```diff
@@ -0,0 +1,22 @@
+ */
+class StatementDescriptor(rawValue: String?) {
+
+    val value: String? = rawValue?.take(MAX_LENGTH)?.split("<", ">", "'", "\"")?.joinToString(REPLACEMENT_CHAR)
```
We probably also should address min (`5`) required size here:

> Contains between 5 and 22 characters, inclusive.

kidinov replied:
> Doesn’t contain any of the following special characters: <, >, \, ' " *.

It looks like that we are missing `*` handling

samiuelson replied:
Interesting, it looks like there are different requirements on Android and iOS related to descriptor validation. The current implementation takes into account iOS. I'm going to change it!

kidinov replied:
* Contains at least one letter (if using a prefix and a suffix, both require at least one letter).
* Contains only Latin characters.

Those too. I am not sure what we should do in these cases. Just replacing it with some Latin chars will make a statement not readable. Maybe it will be better to just pass an empty (null) statement

kidinov replied:
@samiuelson 

> Interesting, it looks like there are different requirements on Android and iOS related to descriptor validation. The current implementation takes into account iOS. 

Yeah, it's either the documentation has changed, or iOS implemented was not fully done back then 🤷‍♂️

samiuelson replied:
I discovered that according to Android SDK `Builder::setStatementDescriptor` [docs](https://stripe.dev/stripe-terminal-android/external/com.stripe.stripeterminal.external.models/-payment-intent-parameters/-builder/set-statement-descriptor.html) there are the following requirements for statement text:
1. Text must be max 22 chars in length.
3. It must not contain the following characters: <, >, ', ", *.
4. It must contain at least one letter.

Luckily the SDK strips non-ASCII chars internally so we don't need to worry about it 😮‍💨.

I've updated the requirements' checks and text processing accordingly.

samiuelson committed:
Include "*" into illegal chars requirement

kidinov reviewed (commented):
Huge thanks for the PR! 

Please take a look into comments I left and let me know what you think

samiuelson committed:
Handle min length requirement

samiuelson committed:
Fix RELEASE-NOTES.txt

samiuelson committed:
Fix RELEASE-NOTES.txt

samiuelson committed:
Update validation rules

Apply validation rules from the following doc: https://stripe.dev/stripe-terminal-android/external/com.stripe.stripeterminal.external.models/-payment-intent-parameters/-builder/set-statement-descriptor.html

 1. Text must be max 22 chars long.
 2. Text must be min 5 chars long.
 3. It must not contain the following characters: <, >, ', ", *.
 4. It must contain at least one letter.

 Note: non-ASCII chars are stripped internally by Stripe.

samiuelson committed:
Merge branch 'trunk' into issue-7586/statement-descriptor-improvement

samiuelson committed:
Fix detekt issue

Detekt complains about line being too long in kdoc section.
We can't break this line because it contains of a long URL.

samiuelson committed:
Merge branch 'trunk' into issue-7586/statement-descriptor-improvement

samiuelson committed:
Merge branch 'trunk' into issue-7586/statement-descriptor-improvement

kidinov commented on libs/cardreader/src/test/java/com/woocommerce/android/cardreader/internal/payments/StatementDescriptorTest.kt:
```diff
@@ -0,0 +1,88 @@
+class StatementDescriptorTest {
+
+    @Test
+    fun `given statement descriptor longer than 22 chars first 22 chars should be returned`() {
```
Sorry, didn't notice this in the first round. We have [a convention](https://github.com/woocommerce/woocommerce-android/blob/trunk/CONVENTION.md#naming-1) how we call tests, and you pretty much follow it, but still small changes required

kidinov reviewed (approved):
LGTM! I would just rename the tests to follow the convention and we are good to merge

samiuelson committed:
Update test method names

samiuelson committed:
Suppress detekt max length rule in StatementDescriptorTest

samiuelson committed:
Merge branch 'trunk' into issue-7586/statement-descriptor-improvement
